storage/namespace.go

```diff
@@ -172,21 +165,63 @@ func (n *dbNamespace) WriteBatch(
 	ctx context.Context,
 	docs []document.Document,
 ) error {
+	if len(docs) == 0 {
+		return nil
+	}
 	builder := n.newSegmentBuilderFn(n.segmentBuilderOpts)
 	builder.AddBatch(docs)
 	rawSegment := builder.Build()
-	newSegment := newDatabaseSegment(n.id, rawSegment, n.segmentOpts)
+	payload := &segmentPayload{
+		namespace: n.id,
+		seg:       rawSegment,
+	}
+	payload.wg.Add(1)
 
-	n.Lock()
+	// Enqueue the segment.
+	n.RLock()
 	if n.closed {
-		n.Unlock()
+		n.RUnlock()
+		rawSegment.Close()
 		return errNamespaceAlreadyClosed
 	}
-	n.unflushed = append(n.unflushed, newSegment)
-	n.sealedByMaxTimeAsc.Set(float64(rawSegment.Metadata().MaxTimeNanos), newSegment)
-	n.Unlock()
+	err := n.flushManager.Enqueue(payload)
+	n.RUnlock()
 
-	return nil
+	if err != nil {
+		rawSegment.Close()
+		return err
+	}
+
+	// Wait for flush to finish if needed.
+	callbackFn := func() error {
+		payload.wg.Wait()
+		if payload.resultErr != nil {
+			rawSegment.Close()
+			return payload.resultErr
+		}
+
+		// On flush success, clear in-memory data (but keep the metadata) and update the list
+		// of searchable segments.
+		rawSegment.ClearAll()
+
+		n.Lock()
+		defer n.Unlock()
+
+		if n.closed {
+			rawSegment.Close()
```

Comment:
hmm, we'll have an odd situation where the segment has already been persisted to disk but not searchable.

Reply:
Yeah this is ok because it's done during namespace close (i.e., database shutdown), so the database will find such segments when the db starts up with proper bootstrapping logic.